### PR TITLE
refactor(types): prevent `ScalarImpl` and `Datum` from incorrectly ordered

### LINKED ISSUE #4272: scalar: introduce Datum struct to have null-last Ord behavior

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Un-impl `PartialOrd` and `Ord` for `ScalarImpl` and `Datum`, and introduce `DefaultPartialOrd` and `DefaultOrd` for them, to avoid incorrectly ordering these types.

This should resolve #4272.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- ~~[ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/util/sort_util.rs [hunk 1]
```diff
@@ -399,17 +405,97 @@ pub fn compare_rows_in_chunk(
...
     rhs: impl ToDatumRef,
     order_type: OrderType,
+) -> Option<Ordering> {
+    #[derive(PartialEq, Eq)]
+    struct OrdScalarRefImpl<'a>(ScalarRefImpl<'a>);
+    impl PartialOrd for OrdScalarRefImpl<'_> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            crate::types::default_partial_cmp_scalar_ref_impl(self.0, other.0)
+        }
+    }
```
Comment (richardchien) [inline src/common/src/util/sort_util.rs:420]: Make `OrdScalarRefImpl` local to this function to avoid abuse.

**Review by soundOfDestiny — APPROVED** [review]
LGTM. Great!

File: src/common/src/types/ordered.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                $((ScalarRefImpl::$variant_name(lhs_inner), ScalarRefImpl::$variant_name(ref rhs_inner)) => Some(lhs_inner.cmp(rhs_inner)),)*
+                _ => None,
+            }
+        }
+    };
+}
+for_all_scalar_variants!(gen_default_partial_cmp_scalar_ref_impl);
+
+/// Variant of [`PartialOrd`] that compares with default order.
+pub trait DefaultPartialOrd {
```
Comment (BugenZhao) [inline src/common/src/types/ordered.rs:41]: I guess we can omit `PartialOrd` for a self-defined `Ord` trait. Currently `PartialOrd` returns `None` only if they're in different types, which is not expected and simply panicking is okay there.

File: src/common/src/types/ordered.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+impl DefaultOrd for DatumRef<'_> {
+    fn default_cmp(&self, other: &DatumRef<'_>) -> Ordering {
+        cmp_datum(*self, *other, OrderType::default())
+    }
+}
+
+/// Wrapper of [`ScalarImpl`] that can be sorted in default order.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, EstimateSize)]
+pub struct OrdScalarImpl {
```
Comment (BugenZhao) [inline src/common/src/types/ordered.rs:103]: What about naming it as `DefaultOrdScalarImpl`? Same for `OrdRow`.

Comment (BugenZhao) [inline src/common/src/types/ordered.rs:103]: Maybe we can have a generic wrapper of `DefaultOrdered`, whose functionality is to delegate `std::cmp::Ord` to `DefaultOrd`. 🤩

```rust
trait Row {}

trait DefaultOrd {
    fn cmp_default(&self, _: &Self) -> std::cmp::Ordering;
}

struct DefaultOrdered<T>(T);

impl<T: DefaultOrd> PartialEq for DefaultOrdered<T> {
    fn eq(&self, _other: &Self) -> bool {
        todo!()
    }
}

impl<T: DefaultOrd> PartialOrd for DefaultOrdered<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.0.cmp_default(&other.0).into()
    }
}

impl<T: DefaultOrd> Eq for DefaultOrdered<T> {}

impl<T: DefaultOrd> Ord for DefaultOrdered<T> {
    fn cmp(&self, _: &Self) -> std::cmp::Ordering {
        std::cmp::Ordering::Equal
    }
}

impl<R: Row> Row for DefaultOrdered<R> {}

type DefaultOrdOwnedRow = DefaultOrdered<OwnedRow>;
type DefaultOrdScalarImpl = DefaultOrdered<ScalarImpl>;
```

Comment (richardchien) [reply src/common/src/types/ordered.rs]: Introduced `DefaultOrdered<T>`, and removed `OrdScalarImpl`, `OrdDatum` and `OrdRow`.

**Review by BugenZhao — APPROVED** [review]
LGTM!

**Pull request merged**